File: ios/Classes/utils/AbstractActionModel.swift
//
//  AbstractActionModel.swift
//  wear3_flutter_plugin
//
//  Created by Mr Nguyen Truong Duy on 12/11/2021.
//

import Flutter
import Foundation
import SwiftProtobuf

class AbstractActionModel<Action: SwiftProtobuf.Message> {
    private var _methodChannel: FlutterMethodChannel?
    public var methodChannelId: String

    func buildAction(data: Data) -> Action? {
        do {
            return try Action(serializedData: data)
        } catch {
            print("buildAction error: \(error)")
        }
        return nil
    }

    func handleAction(action _: Action, result: @escaping FlutterResult) {
        preconditionFailure("This method must be overridden")
    }

    func dispose() {
        _methodChannel?.setMethodCallHandler(nil)
        ChannelUtil.shared.cleanUpMethodChannel(for: methodChannelId)
    }

    init(methodChannelId: String? = nil, eventChannelId: String? = nil) {
        self.methodChannelId = methodChannelId ?? UUID().uuidString
        _methodChannel = ChannelUtil.shared.createMethodChannel(id: self.methodChannelId)
        _methodChannel?.setMethodCallHandler { call, result in
            if call.method == "action" {
                let args = call.arguments as! [String: Any]
                let actionData = args["data"] as! FlutterStandardTypedData
                let actionModel = self.buildAction(data: actionData.data)
                if actionModel != nil {
                    print("[iOS] [\(String(describing: type(of: self)))] action: \(try! actionModel!.jsonString())")
                    self.handleAction(action: actionModel!, result: result)
                }
            } else if(call.method == "dispose") {
                print("[iOS] [\(String(describing: type(of: self)))] action: dispose")
                result(nil)
                self.dispose()
            }
        }
    }
}

File: android/src/main/kotlin/com/example/plugin_using_protobuf/utils/AbstractActionModel.kt
package com.example.plugin_using_protobuf.utils

import com.fossil.wear3_flutter_plugin.wear3_flutter_plugin.utils.ChannelUtil
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.*

abstract class AbstractActionModel<A>(
    val binaryMessenger: BinaryMessenger,
    val methodChannelId: String = UUID.randomUUID().toString()
) : MethodChannel.MethodCallHandler where A : com.google.protobuf.MessageLite {
    init {
        this.setup()
    }

    private lateinit var _methodChannel: MethodChannel


    abstract fun buildAction(data: ByteArray): A

    abstract fun handleAction(action: A, result: MethodChannel.Result)

    override fun onMethodCall(method: MethodCall, result: MethodChannel.Result) {
        if (method.method == "action") {
            val actionData = method.arguments as HashMap<*, *>
            val data = actionData["data"] as ByteArray
            val actionModel = this.buildAction(data)
            this.handleAction(actionModel, result)
        } else if (method.method == "dispose") {
            result.success(null)
            this.dispose()
        }
    }

    private fun setup() {
        this._methodChannel = ChannelUtil.INSTANCE.createMethodChannel(methodChannelId, binaryMessenger)
        this._methodChannel.setMethodCallHandler(this)
    }

    open fun dispose() {
        this._methodChannel.setMethodCallHandler(null)
    }
}